keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,509 @@
+This controller will monitor changes in node labels, and try to evict pods if their `requiredDuringSchedulingRequiredDuringExecution` node affinity are no longer met.
+If eviction fails, the pods that failed to be evicted will be rechecked for violations of the `requiredDuringSchedulingRequiredDuringExecution` affinity during the next node label update. 
+If, in the next eviction round, they still violate the `requiredDuringSchedulingRequiredDuringExecution` affinity, they will be attempted for eviction again.
+This controller will respect PDB during the eviction.
+
+We'll start using `Eviction API with PDB` to evict pods during alpha stage for simplicity, 
+and transfer to `Evacuation API` in beta stage.
+
+## Motivation
+
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `node-affinity-eviction` controller to ensure pods being evicted if the selectors are no longer met at runtime.
+- DaemonSet controller will take the `RequiredDuringSchedulingRequiredDuringExecution` into consideration, 
+when recalculating the eligible nodes where the DaemonSet pods can run.
+
+### Non-Goals
+
+- This KEP does not aim to introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type and **podAntiAffinity** type.
+
+## Proposal
+- Implement the `requiredDuringSchedulingRequiredDuringExecution` node affinity type.
+- Update the kube-scheduler, kube-controller-manager and kube-apiserver.
+- Define the behavior, syntax, and use cases in documentation.
+- Facilitate testing, adoption, and maintenance of this feature.
+
+### User Stories (Optional)
+
+#### manage multi tenant cluster with dedicated nodes
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows I want to add new nodes only for user B while leaving the current nodes only for user A.
+I want to migrate all pods available to user B to new nodes, and I want to ensure HA of my services during the migration.
+
+Without `node-affinity-eviction`, I have to remove "userB=allow" label of node, and delete the pods manually. Also I can't use taints because they don't respect PDBs.
+With `node-affinity-eviction`, I can simply delete the "userB=allow" label from the existing nodes 
+to re-schedule all pods of user B to these new nodes.
+
+#### handle data locality pods when data is corrupted
+
+I have one application whose pods need to be co-located with some specific data stored in some particular nodes. 
+When one node's storage is corrupted, the pod on this node will provide some inaccurate services.
+So I want it to be scheduled to another special node when data is detected to be corrupted.
```

Comment:
In this case, we shouldn't respect PDB, but this feature intends to respect PDB.

So, should this story be removed?

Reply:
This can be an example where we don't need to adhere to PDB, and I am inclined to retain it. While taints can achieve this in most cases, we mentioned  in Alternatives section that this offers greater flexibility compared to taints.

Reply:
Whenever you're using evictions API you're automatically respecting PDB, which is one of the main premises of this document. So I'm with Aldo on this one, that we should not provide counterexamples.